Clean up and refactoring

j0sh opened the pull request:
Good to go for review. Tests fail at the moment but those will be fixed up during the week; the code seems functionally OK in conjunction with https://github.com/livepeer/go-livepeer/pull/507 and https://github.com/livepeer/go-livepeer/pull/508.

* [x] Clean up basicnet
* [x] Fix tests
* [x] Split broadcaster code from livepeernode
* [x] Split transcoder code from livepeernode
* [x] Move implementation of the Orchestrator interface outside RPC
* [x] Move implementation of the Broadcaster interface outside RPC
* [x] Move transcoder init routine outside mediaserver.
* [x] Add a (stub) local video network

The empty interfaces for TranscoderInfo are remedied in https://github.com/livepeer/go-livepeer/pull/508/commits/229f87a80d490ee8fa457666b9664c401585e201

Tests are fixed in https://github.com/livepeer/go-livepeer/pull/511

j0sh committed:
Code cleanup.

j0sh committed:
Split transcoder-specific code into its own file.

j0sh committed:
Split broadcaster-specific code into its own file.

j0sh committed:
transcoder: Return a value after transcoding.

j0sh committed:
mediaserver: Share HTTP mux between LPMS and RPC.

j0sh committed:
cli: Make server use non-default mux.

Because we can elect to use the default mux in LPMS sometimes.

j0sh committed:
livepeer: Make LPMS HTTP address configurable.

j0sh committed:
videonetwork: Add a local playlist cache.

j0sh committed:
rpc: Generate code for transcoder responses.

j0sh committed:
rpc: Send transcoded results back to client.

j0sh commented:
Note that in https://github.com/livepeer/go-livepeer/pull/506/commits/2059fe0531443266135659e0c1f9c65be7cc8403 we introduce a "local video network" which is essentially a cache for the master playlist. This is structurally very similar to what we already have in the VideoCache.

Wondering a few things:
* Can we merge the VideoNetwork and the VideoCache to flatten this a bit? There isn't a strong concept of a 'video network' anymore. Alternatively, we could subsume VideoCache into the VideoNetwork interface, which might give us some room for simplification.
* We will probably want to parameterize a "video cache" (or, perhaps equivalently, a "video network") by the object store. Note that each job may have several object stores. Will need to consider how to manage this.

Thoughts?

j0sh committed:
transcoder: Simplify the internal API.

j0sh committed:
claimmanager: Return transcoded data hash.

j0sh committed:
monitor: Accommodate new transcode response format.

j0sh committed:
broadcaster: Download segments from the transcoder.

j0sh committed:
broadcaster: Move RPC interface impl into broadcaster.go

Temporarily set the TranscoderInfo to an empty interface
until we refactor the protobufs into the `net` package.

j0sh committed:
transcoder: Move RPC interface impl into transcoder.go

j0sh committed:
transcoder: Move initialization routine outside of mediaserver.

j0sh committed:
livepeer: Split Service URI config from RPC listener.

j0sh committed:
proto: Move protobufs definitions into net.

Allows us to use the data types outside the server package.

j0sh committed:
proto: Update references to new proto location.

j0sh committed:
vendor: Update LPMS to d8a9224.

ericxtang commented on core/broadcaster.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+//CreateTranscodeJob creates the on-chain transcode job.
+func (n *LivepeerNode) CreateTranscodeJob(strmID StreamID, profiles []ffmpeg.VideoProfile, price *big.Int) (*ethTypes.Job, error) {
```
Having `LivepeerNode` methods in `broadcaster.go` feels a little misleading.  I think we should either keep the `CreateTranscodeJob` method in the `livepeernode.go` file, or create a `livepeernode` package that includes multiple files that only contain `LivepeerNode`-specific methods (for example, we can have `transcodeJob.go` in `/core/livepeernode/`).  It feels like the "job creation" responsibility should belong to `LivepeerNode`.

j0sh replied:
Part of the idea was to keep functionally related code together, and unrelated code separate. Given the current setup in `core` :

* transcoder.go : Code that is meant to be called only by the transcoder / orchestrator. That's `TranscodeSegment` related and the RPC interface implementation.
* broadcaster.go : Code that is meant to be called only by the broadcaster. `CreateTranscodeJob` is the major one here (transcoders don't create jobs), in addition to the RPC interface impl.
* `livepeernode.go` - Shared code common to both node types. There is very little. Looking at what we have left in livepeernode, `GetClaimManager` probably should also be moved to the transcoder/orchestrator. Many of the LivepeerNode fields are also transcoder-specific.

This separation could be taken further. We can remove `LivepeerNode` as a standalone entity, and create a  `BroadcasterNode` and `TranscoderNode` .  The livepeernode struct can be kept, and the broadcaster/transcoder can inherit common fields. We might have to tweak StartEthServices/StopEthServices (the only shared functions right now) but that is relatively minor.

Personally, I feel that the implicit context given by the filename provides enough of a clue as to where/how the code should be used, regardless of the underlying 'base struct'. We can go too far in subdividing code, eg files implementing only one function such as `transcodeJob.go`. But that boils down to preferences.

ericxtang replied:
I see.  That seems logical - can we add this logical layout as a comment in `livepeernode.go` for documentation?

j0sh replied:
Yep, comments added here https://github.com/livepeer/go-livepeer/pull/512/commits/84e0261776b74d862879ec4fd8a54ee336228896

ericxtang commented on core/localnetwork.go:
```diff
@@ -0,0 +1,59 @@
+package core
```
The current implementation of `VideoCache` (`BasicVideoCache`) has a tight integration with basicnet.  I had the assumption that `VideoCache` implicitly has a data source, so it only need to have `Get` and `Evict` functions, and never need to have `Update` functions (`Get` will automatically check from the data source).

What about expanding the `VideoCache` interface to add `Update` functions?  We can rename `BasicVideoCache` to `BasicnetVideoCache`, and it will implement something like:
```
func UpdateMasterPlaylist(manifestID string, mpl *m3u8.MasterPlaylist) error {
  return errors.New("Cannot Update in BasicnetVideoCache")
}
```
and move `LocalNetwork` to `LocalVideoCache`, which can implement the `Evict` and `Update` functions.

j0sh replied:
We can add `Update` functions to the interface, however I'm not really sure if we need a formalized interface so much anymore now that basicnet is gone. We'll see how things look after the VideoNetwork and the VideoCache/Source are combined.

However, at some point we will likely need to parameterize the VideoSource by the object store. That part can wait though.

ericxtang commented on core/transcoder.go:
```diff
@@ -0,0 +1,326 @@
+
+type SegmentChan chan *SegChanData
+
+func (n *LivepeerNode) getSegmentChan(job *ethTypes.Job) (SegmentChan, error) {
```
Feels like this file is really about orchestrators.  How about renaming it to `orchestrator.go`?  Also, similar comment here to the case above, where I think the `LivepeerNode` methods should be moved to where `livepeernode.go` lives to avoid confusion.  Imagine a new developer coming into the codebase - she might get disoriented.

j0sh replied:
Sure. Most of the interesting bits in this file involve transcoding, but we can rename to `orchestrator.go` given future plans.

j0sh replied:
transcoder.go renamed to orchestrator.go here https://github.com/livepeer/go-livepeer/pull/512/commits/7a009aeedd8319be4b96ec7f687e3ffe956b5132

ericxtang commented on net/interface.go:
```diff
@@ -10,21 +9,14 @@ import (
 )
 
 //VideoNetwork describes the interface for a Livepeer node network-layer library.
 type VideoNetwork interface {
```
It's starting to feel like `VideoNetwork` and `VideoCache` has some similarities.  On a high level, I feel `VideoCache` is probably a misnomer for what it is today.  Maybe `VideoSource` is more appropriate.  If that's the case, maybe a `VideoNetwork` implements `VideoSource` + extra network-related methods like `GetNodeStatus`.  `VidoeSource` also plays well with a future object store implementation.

j0sh replied:
+1 for `VideoSource`. We can flatten the VideoNetwork into VideoCache(Source) and see where that takes us.

j0sh replied:
Flattening [here](https://github.com/livepeer/go-livepeer/pull/512/commits/2da29308070528f6f4c151292f4b264d26e0cfd3) and file rename [here](https://github.com/livepeer/go-livepeer/pull/512/commits/0a0e4e6f4093eb74550d1b538cfe072d79de1241)

Note that a side effect is we don't really a need for a VideoNetwork interface anymore, so that was removed.

ericxtang reviewed (commented):
This is awesome - I feel like we took a big step forward in simplifying the code here.  I left some architectural feedback.

j0sh committed:
core: Fix tests.

j0sh committed:
server: Fix tests.

j0sh committed:
eth: Fix tests.

j0sh committed:
livepeer: Merge rtmpIp and rtmpPort into rtmpAddr.

j0sh committed:
rpc: Privatize getTranscoder.

j0sh committed:
transcoder: Cosmetic renames.

j0sh committed:
videocache: Flatten VideoNetwork.

This also leaves VideoNetwork with no significant purpose, so fold
the remaining into VideoCache and remove VideoNetwork.

j0sh committed:
core: Additional test cleanups.

j0sh committed:
transcoder: Move GetClaimManager into transcoder.go.

Also document the logical layout of the code within LivepeerNode.

j0sh committed:
orchestrator: Rename file.

j0sh committed:
videosource: Rename from VideoCache.

j0sh committed:
rpc: Fix crash on non-200 response.

j0sh committed:
core: Remove the notion of a NodeID.

Also add some tests around StreamID / ManifestID.

j0sh committed:
livepeer: Switch default ports around.

HTTP/RPC becomes 8935
CLI becomes 7935

j0sh committed:
livepeer: Use blockchain ServiceURI rather than locally inferred URI.

darkdarkdragon committed:
Using blocking connection to the rpc server - this will give us clear error in case transcoder is down

j0sh committed:
core: Length checks for ManifestID/StreamID.

Prevents crashes on malformed input.

j0sh committed:
jobwatcher: Make log parsing more robust.

Prevents crashes with malformed StreamIDs.

j0sh committed:
jobwatcher: Retry the assigned transcoder lookup.

More robust against sporadic Eth RPC failures.

j0sh committed:
rpc: Misc fixes.

j0sh committed:
README: Update.

darkdarkdragon commented on server/mediaserver.go:
```diff
@@ -595,7 +530,6 @@ func (s *LivepeerServer) startHlsUnsubscribeWorker(limit time.Duration, freq tim
 			if time.Since(t) > limit {
 				glog.V(common.SHORT).Infof("Inactive HLS Stream %v - unsubscribing", sid)
-				s.LivepeerNode.UnsubscribeFromNetwork(sid)
 				delete(s.hlsSubTimer, sid)
```
@ericxtang looks like whole `startHlsUnsubscribeWorker` function along with `hlsSubTimer` field should be removed - it doesn't do anything useful now

ericxtang replied:
Good point @darkdarkdragon.  I'll make the change.

yondonfu replied:
Removed in 9737e7057b37880cd1f7d04324bbfca01c6adc96

ericxtang committed:
Merge pull request #513 from livepeer/ja/nodeid

Remove Node ID

yondonfu committed:
Merge pull request #516 from livepeer/it/errors_logging

Better errors reporting

yondonfu committed:
Merge pull request #512 from livepeer/ja/v2followup

Follow up changes to v2 feedback

yondonfu committed:
Merge pull request #511 from livepeer/ja/fixtests

Fix tests

yondonfu committed:
Merge pull request #508 from livepeer/ja/tcresponse

Structured responses from transcoder

yondonfu committed:
Merge pull request #507 from livepeer/ja/lpmsmux

Split up LPMS HTTP listener from CLI

yondonfu committed:
mediaserver: Remove startHlsUnsubscribeWorker and hlsSubTimer because they no longer do anything useful

yondonfu reviewed (approved):
🚢
